sycl/include/sycl/ext/oneapi/kernel_properties/properties.hpp

```diff
@@ -351,6 +350,78 @@ struct HasKernelPropertiesGetMethod<T,
       decltype(std::declval<T>().get(std::declval<properties_tag>()));
 };
 
+// Trait for property compile-time meta names and values.
+template <typename PropertyT> struct WGSizePropertyMetaInfo {
+  static constexpr std::array<size_t, 0> WGSize = {};
+  static constexpr size_t LinearSize = 0;
+};
+
+template <size_t Dim0, size_t... Dims>
+struct WGSizePropertyMetaInfo<work_group_size_key::value_t<Dim0, Dims...>> {
+  static constexpr std::array<size_t, sizeof...(Dims) + 1> WGSize = {Dim0,
+                                                                     Dims...};
+  static constexpr size_t LinearSize = (Dim0 * ... * Dims);
+};
+
+template <size_t Dim0, size_t... Dims>
+struct WGSizePropertyMetaInfo<max_work_group_size_key::value_t<Dim0, Dims...>> {
+  static constexpr std::array<size_t, sizeof...(Dims) + 1> WGSize = {Dim0,
+                                                                     Dims...};
+  static constexpr size_t LinearSize = (Dim0 * ... * Dims);
+};
+
+// Get the value of a work-group size related property from a property list
+template <typename PropKey, typename PropertiesT>
+struct GetWGPropertyFromPropList {};
+
+template <typename PropKey, typename... PropertiesT>
+struct GetWGPropertyFromPropList<PropKey, std::tuple<PropertiesT...>> {
+  using prop_val_t = std::conditional_t<
+      ContainsProperty<PropKey, std::tuple<PropertiesT...>>::value,
+      typename FindCompileTimePropertyValueType<
+          PropKey, std::tuple<PropertiesT...>>::type,
+      void>;
+  static constexpr auto WGSize =
+      WGSizePropertyMetaInfo<std::remove_const_t<prop_val_t>>::WGSize;
+  static constexpr size_t LinearSize =
+      WGSizePropertyMetaInfo<std::remove_const_t<prop_val_t>>::LinearSize;
+};
+
+// If work_group_size and max_work_group_size coexist, check that the
+// dimensionality matches and that the required work-group size doesn't
+// trivially exceed the maximum size.
+template <typename Properties>
+struct ConflictingProperties<max_work_group_size_key, Properties>
+    : std::false_type {
+  using WGSizeVal = GetWGPropertyFromPropList<work_group_size_key, Properties>;
+  using MaxWGSizeVal =
+      GetWGPropertyFromPropList<max_work_group_size_key, Properties>;
+  static constexpr size_t Dims = WGSizeVal::WGSize.size();
+  static_assert(
+      Dims == 0 || Dims == MaxWGSizeVal::WGSize.size(),
```

Comment:
It took me a moment to realize that `Dims == 0` is for the case where `WGSizeVal` doesn't exist in the list. Definitely clever, but would you mind adding a comment to make it a little easier for the next time someone comes by this?

Reply:
Yeah, good idea. Even took me a while to remember what I'd done.